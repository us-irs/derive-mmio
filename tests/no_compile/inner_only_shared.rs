#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct UartBank {
    // this is read-write by default
    data: u32,
}

#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct Uart {
    #[mmio(inner)]
    bank_0: UartBank,
}

fn main() {
    let mut uart = Uart {
        bank_0: UartBank {
            data: 0x1,
        },
    };
    // Safety: We're pointing at a real object
    let mmio_uart = unsafe { Uart::new_mmio(core::ptr::addr_of_mut!(uart)) };

    // Shared inner block, can only use shared API.
    let bank0_shared = mmio_uart.bank_0_shared();
    bank0_shared.write_data(0x2);
}
