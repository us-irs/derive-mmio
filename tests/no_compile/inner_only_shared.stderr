error[E0596]: cannot borrow data in dereference of `SharedInner<MmioUartBank<'_>>` as mutable
  --> tests/no_compile/inner_only_shared.rs:26:5
   |
26 |     bank0_shared.write_data(0x2);
   |     ^^^^^^^^^^^^ cannot borrow as mutable
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `SharedInner<MmioUartBank<'_>>`
